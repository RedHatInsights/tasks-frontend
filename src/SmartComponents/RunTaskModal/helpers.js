export const findParameterByKey = (parameters, key) => {
  return parameters.find((param) => param.key === key);
};

export const toBool = (value) => {
  return value?.toLowerCase() === 'true' || value === '1' || value === 1;
};

const isTextbox = (parameter) => (parameter?.values || []).length === 0;

const isCheckbox = (parameter) =>
  parameter.values.length === 2 &&
  (parameter.values.every((value) =>
    ['true', 'false'].includes(value.toLowerCase())
  ) ||
    parameter.values.every((value) => ['0', '1'].includes(value)));

const isMultiSelect = (parameter) => parameter.multi_valued;

// This will likely change when the Tasks backend supports Data Driven Forms
export const getInputParameterType = (parameter) => {
  switch (true) {
    case isCheckbox(parameter):
      return 'checkbox';
    case isTextbox(parameter):
      return 'textbox';
    case isMultiSelect(parameter):
      return 'multiselect';
    default:
      return 'dropdown';
  }
};
