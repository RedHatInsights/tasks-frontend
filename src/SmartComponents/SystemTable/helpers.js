import { ALL_SYSTEMS_VALUE } from './constants';

const buildSortString = (orderBy, orderDirection) => {
  let sortString = orderBy ? '&sort=' : '';
  let direction = '';
  if (orderDirection === 'DESC') {
    direction = '-';
  }
  let order =
    (orderBy === 'updated' && 'last_seen') ||
    (orderBy === 'os_version' && 'os') ||
    orderBy;

  return `${sortString}${direction}${order}`;
};

const buildOsFilterString = (osFilter) =>
  Object.entries(osFilter)
    .map(([majorOsVersion, minorOsVersions]) =>
      Object.entries(minorOsVersions)
        .map(([minorOsKey, value]) => {
          if (minorOsKey !== majorOsVersion && value === true) {
            const osArray = minorOsKey.split('-');
            const minorVersion = osArray.slice(
              osArray.length - 1,
              osArray.length
            );
            const osName = osArray.slice(0, osArray.length - 2).join(' ');

            return minorVersion && `${osName}|${minorVersion}`;
          }
        })
        .filter((v) => !!v)
        .join(',')
    )
    .filter((v) => !!v)
    .join(',');

const buildFilterString = (filters = {}) => {
  let osFiltersString = '';
  let groupsFilterString = '';
  let displayNameFilter = filters.hostnameOrId
    ? `&display_name=${filters.hostnameOrId}`
    : '';

  if (filters.osFilter) {
    const builtOsFilterString = buildOsFilterString(filters.osFilter);
    osFiltersString += builtOsFilterString
      ? '&operating_system=' + builtOsFilterString
      : '';
  }

  filters.hostGroupFilter?.forEach((group) => {
    groupsFilterString += `&groups=${group}`;
  });

  return `${displayNameFilter}${osFiltersString}${groupsFilterString}`;
};

const buildTagsFilterString = (tags, filters) => {
  let tagsFilterString = tags?.length ? '&tags=' : '';
  let globalTagsFilterString = '';
  let tagFiltersString = '';
  tags?.forEach((tag, index) => {
    index ? (globalTagsFilterString += ',') : (globalTagsFilterString += '');
    globalTagsFilterString += `${tag}`;
  });
  filters.tagFilters?.forEach((tag, index) => {
    index === 0 && globalTagsFilterString === ''
      ? (tagFiltersString += '')
      : (tagFiltersString += ',');
    tag.values.forEach(
      (value) => (tagFiltersString += `${tag.category}/${value.name}`)
    );
  });

  return `${tagsFilterString}${globalTagsFilterString}${tagFiltersString}`;
};

const buildWorkloadFiltersString = (filters) => {
  let workloadFilterString = '';
  for (const [key, value] of Object.entries(filters)) {
    workloadFilterString += `&${key}=${value}`;
  }

  return encodeURI(workloadFilterString);
};

const buildEligibilityFilterString = ({ filters }) => {
  return filters[0]?.chips[0]?.value === ALL_SYSTEMS_VALUE
    ? '&all_systems=true'
    : '';
};

// TODO this should be based on a URLSearchParams object and use its toString() function
export const buildFilterSortString = (
  limit,
  offset,
  orderBy,
  orderDirection,
  filters,
  tags,
  workloadFilters,
  activeFiltersConfig
) => {
  let limitOffsetString = `limit=${limit}&offset=${offset}`;
  let sortString = buildSortString(orderBy, orderDirection);
  let filterString = buildFilterString(filters);
  let tagsFilterString = buildTagsFilterString(tags, filters);
  let workloadFilterString = buildWorkloadFiltersString(workloadFilters);
  let eligibilityFilterString =
    buildEligibilityFilterString(activeFiltersConfig);
  return `?${limitOffsetString}${sortString}${filterString}${tagsFilterString}${workloadFilterString}${eligibilityFilterString}`;
};

export const findCheckedValue = (total, selected) => {
  if (selected === total && total > 0) {
    return true;
  } else if (selected > 0 && selected < total) {
    return null;
  } else {
    return false;
  }
};

export const findFilterData = (optionName, options) => {
  return options.find((item) => item.label === optionName);
};
