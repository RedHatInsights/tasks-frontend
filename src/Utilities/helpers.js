import React from 'react';
import moment from 'moment';
import {
  Skeleton,
  SkeletonSize,
} from '@redhat-cloud-services/frontend-components/Skeleton';

export const renderRunDateTime = (time) => {
  if (time === 'loading') {
    return <Skeleton size={SkeletonSize.md} />;
  } else {
    return moment.utc(time).format('DD MMM YYYY, HH:mm UTC');
  }
};

export const getTimeDiff = ([start, end, status]) => {
  if (start === 'loading') {
    return <Skeleton size={SkeletonSize.md} />;
  } else {
    return status === 'Running' || !end
      ? '-'
      : `${renderRunDateTime(end)} (${moment
          .duration(
            moment(renderRunDateTime(end), 'DD MMM YYYY, HH:mm').diff(
              moment(renderRunDateTime(start), 'DD MMM YYYY HH:mm')
            )
          )
          .asMinutes()} min)`;
  }
};

export const uniq = (collection) => [...new Set(collection)];

/*eslint-disable react/display-name*/
export const renderColumnComponent =
  (Component, props) => (_data, _id, entity) =>
    <Component {...entity} {...props} />;
/*eslint-enable react/display-name*/

const getSortable = (property, item) => {
  if (typeof property === 'function') {
    return property(item);
  } else {
    return item[property];
  }
};

export const orderArrayByProp = (property, objects, direction) =>
  objects.sort((a, b) => {
    if (direction === 'asc') {
      if (typeof a[property] === 'number') {
        return getSortable(property, a) - getSortable(property, b);
      } else {
        return String(getSortable(property, a)).localeCompare(
          String(getSortable(property, b))
        );
      }
    } else {
      if (typeof a[property] === 'number') {
        return -getSortable(property, a) + getSortable(property, b);
      } else {
        return -String(getSortable(property, a)).localeCompare(
          String(getSortable(property, b))
        );
      }
    }
  });

export const getProperty = (obj, path, fallback) => {
  const parts = path.split('.');
  const key = parts.shift();
  if (typeof obj[key] !== 'undefined') {
    return parts.length > 0
      ? getProperty(obj[key], parts.join('.'), fallback)
      : obj[key];
  }

  return fallback;
};

export const camelCase = (string) => {
  return string
    .split(/[-_\W]+/g)
    .map((string) => string.trim())
    .map((string) => string[0].toUpperCase() + string.substring(1))
    .join('');
};
