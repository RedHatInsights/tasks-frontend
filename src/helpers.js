import React from 'react';
import { Tooltip } from '@patternfly/react-core';

export const createSystemLink = (id, name, keyData, isBetaEnv) => (
  <a
    rel="noreferrer"
    target="_blank"
    key={keyData}
    href={
      isBetaEnv
        ? `/preview/insights/inventory/${id}`
        : `/insights/inventory/${id}`
    }
  >
    {name}
  </a>
);

export const createEligibilityTooltip = (eligibility) => {
  return eligibility.tooltip ? (
    <Tooltip content={eligibility.tooltip}>
      <span>{eligibility.title}</span>
    </Tooltip>
  ) : (
    <span>{eligibility.title}</span>
  );
};
