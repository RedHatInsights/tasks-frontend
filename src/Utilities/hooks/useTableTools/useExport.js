import { downloadFile } from '@redhat-cloud-services/frontend-components-utilities/helpers';
import { camelCase, getProperty } from '../../helpers';
import { encodeCsvCell } from './reportParser';

const CSV_FILE_PREFIX = 'tasks-export';
const CSV_DELIMITER = ',';

const filename = () => CSV_FILE_PREFIX + '-' + new Date().toISOString();

const textForCell = (row, column) => {
  const { exportKey, renderExport } = column;
  let cell = exportKey ? getProperty(row, exportKey) : row;
  if (renderExport) {
    return renderExport(cell);
  } else {
    return cell;
  }
};

export const csvForItems = ({ items, columns }) => {
  const header = columns
    .map((column) => column.original || column.title)
    .join(CSV_DELIMITER);
  const csvRows = [
    header,
    ...items.map((row) =>
      columns
        .map((column) => encodeCsvCell(textForCell(row, column)))
        .join(CSV_DELIMITER)
    ),
  ];

  return csvRows.join('\n');
};

export const jsonForItems = ({ items, columns }) => {
  const result = items.map((row) =>
    columns.reduce((object, column) => {
      const key = camelCase(column.original || column.title);
      const value = textForCell(row, column);

      object[key] = value;
      return object;
    }, {})
  );

  return JSON.stringify(result);
};

const callCallback = (callback, ...args) => callback && callback(...args);

const useExport = ({
  exporter,
  columns = [],
  isDisabled = false,
  onStart,
  onComplete,
  onError,
}) => {
  const exportableColumns = columns.filter(
    (column) =>
      column.export !== false && (column.exportKey || column.renderExport)
  );
  const exportWithFormat = async (format) => {
    callCallback(onStart);
    const items = await exporter()
      .then((items) => {
        callCallback(onComplete, items);
        return items;
      })
      .catch((error) => callCallback(onError, error));

    const formater = format === 'csv' ? csvForItems : jsonForItems;

    if (items) {
      downloadFile(
        formater({
          items,
          columns: exportableColumns,
        }),
        filename(),
        format
      );
    } else {
      console.info('No items returned for export');
    }
  };

  return {
    toolbarProps: {
      exportConfig: {
        isDisabled,
        onSelect: (_, format) => exportWithFormat(format),
      },
    },
  };
};

export const useExportWithItems = (items, columns, options = {}) => {
  const exportEnabled = options?.exportable;
  const {
    columns: exportableColumns,
    onStart,
    onComplete,
    prepareItems,
    extraExportColumns = [],
  } = typeof options.exportable === 'object' ? options.exportable : {};

  const exportableSelectedColumns = (exportableColumns || columns).filter(
    (column) => columns.includes(column)
  );

  const exportProps = useExport({
    exporter: () => Promise.resolve(prepareItems ? prepareItems(items) : items),
    columns: exportableSelectedColumns.concat(extraExportColumns),
    isDisabled: items.length === 0,
    onStart,
    onComplete,
  });

  return exportEnabled ? exportProps : {};
};

export default useExport;
