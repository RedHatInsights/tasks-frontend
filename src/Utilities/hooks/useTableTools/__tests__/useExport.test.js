import { act, renderHook } from '@testing-library/react';
import items from './__fixtures__/items.fixtures';
import useExport, {
  jsonForItems,
  csvForItems,
  useExportWithItems,
} from '../useExport';
import {
  fixturesExtendedReport,
  fixturesPlainReport,
} from './__fixtures__/jobResultsItems.fixtures';
import {
  jobCompleteColumnsFixtures,
  taskColumnsFixtures,
} from './__fixtures__/columns.fixtures';
import get from 'lodash/get';
import { waitFor } from '@testing-library/react';
import { downloadFile } from '@redhat-cloud-services/frontend-components-utilities/helpers';

jest.mock('@redhat-cloud-services/frontend-components-utilities/helpers');

describe('useExport', () => {
  let workingExporter = jest.fn(() => Promise.resolve(items));
  let defaultOptions;

  beforeEach(() => {
    defaultOptions = {
      columns: taskColumnsFixtures,
    };
  });

  it('returns an export config toolbar config', () => {
    defaultOptions.exporter = workingExporter;
    const { result } = renderHook(() => useExport(defaultOptions));
    expect(result.current.toolbarProps.exportConfig).toBeDefined();
    expect(result.current).toMatchSnapshot();
  });

  it('returns an export config toolbar config', () => {
    defaultOptions.exporter = workingExporter;
    const { result } = renderHook(() =>
      useExport({
        ...defaultOptions,
        isDisabled: true,
      })
    );
    expect(result.current.toolbarProps.exportConfig.isDisabled).toBe(true);
  });

  it('calls the exporter via onSelect', () => {
    defaultOptions.exporter = workingExporter;
    const { result } = renderHook(() => useExport(defaultOptions));

    act(() => {
      result.current.toolbarProps.exportConfig.onSelect(null, 'csv');
    });

    expect(defaultOptions.exporter).toHaveBeenCalled();

    act(() => {
      result.current.toolbarProps.exportConfig.onSelect(null, 'json');
    });

    expect(defaultOptions.exporter).toHaveBeenCalled();
  });
});

describe('jsonForItems', () => {
  it('returns an json export of items', () => {
    const result = jsonForItems({
      columns: taskColumnsFixtures,
      items: items,
    });

    expect(result).toMatchSnapshot();
  });
});

describe('useExportWithItems', () => {
  it('calls prepareItems when export selected', () => {
    const prepareItemsMock = jest.fn();
    const { result } = renderHook(() =>
      useExportWithItems([], [], {
        exportable: { prepareItems: prepareItemsMock },
      })
    );

    expect(prepareItemsMock).not.toBeCalled();
    result.current.toolbarProps.exportConfig.onSelect();
    expect(prepareItemsMock).toBeCalled();
  });

  it('should add extra columns to the report', async () => {
    const { result } = renderHook(() =>
      useExportWithItems(
        [
          {
            ...fixturesExtendedReport,
            test_key1: 'success',
            test_key2: ['some', 'array', 1],
          },
        ],
        jobCompleteColumnsFixtures,
        {
          exportable: {
            extraExportColumns: [
              {
                title: 'Test 1',
                renderExport: (job) => get(job, 'test_key1', ''),
              },
              {
                title: 'Test 2',
                renderExport: (job) => get(job, 'test_key2', ''),
              },
            ],
          },
        }
      )
    );

    result.current.toolbarProps.exportConfig.onSelect(undefined, 'csv');
    await waitFor(() => {
      expect(downloadFile).toBeCalledWith(
        'System name,Status,Message,Test 1,Test 2\n"System deleted","Success","Completed","success",""some","array",1"',
        expect.anything(),
        'csv'
      );
    });
  });
});

describe('csvForItems', () => {
  it('returns some example csv export of items', () => {
    const result = csvForItems({
      columns: taskColumnsFixtures,
      items: items,
    });

    expect(result).toMatchSnapshot();
  });

  it('returns an csv export of items for complete job', () => {
    const result = csvForItems({
      columns: jobCompleteColumnsFixtures,
      items: [fixturesPlainReport],
    });

    expect(result).toMatchSnapshot();
  });

  it('adds extra double-quote if contained in cell', () => {
    const result = csvForItems({
      columns: taskColumnsFixtures,
      items: items,
    });

    expect(result).toContain('hehe"",""break');
  });

  it('does not escape new line', () => {
    const result = csvForItems({
      columns: taskColumnsFixtures,
      items,
    });

    expect(result).toContain('2022-04\n57\n');
  });
});
