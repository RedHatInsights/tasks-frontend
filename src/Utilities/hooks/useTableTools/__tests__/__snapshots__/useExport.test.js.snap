// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csvForItems returns an csv export of items for complete job 1`] = `
"System name,Status,Message
"host_abc","Success","Completed""
`;

exports[`csvForItems returns some example csv export of items 1`] = `
"Task,Systems,Run date/time
"taskA",10,"2022-04-20T10:10:00"
"taskB",5,"2022-04-21T10:10:00"
"special one",42,"2022-04
57
hehe"",""break!"
"another one with unexpected array",10,"42,"abc"""
`;

exports[`jsonForItems returns an json export of items 1`] = `"[{"Task":"taskA","Systems":10,"RunDateTime":"2022-04-20T10:10:00"},{"Task":"taskB","Systems":5,"RunDateTime":"2022-04-21T10:10:00"},{"Task":"special one","Systems":42,"RunDateTime":"2022-04\\n57\\nhehe\\",\\"break!"},{"Task":"another one with unexpected array","Systems":10,"RunDateTime":[42,"abc"]}]"`;

exports[`useExport returns an export config toolbar config 1`] = `
{
  "toolbarProps": {
    "exportConfig": {
      "isDisabled": false,
      "onSelect": [Function],
    },
  },
}
`;
